core/activity/src/provider/service.rs

```diff
@@ -350,31 +406,41 @@ async fn monitor_activity(
                 break;
             }
 
-            let dt = (Utc::now().timestamp() - usage.timestamp) as f64;
-            if dt > limit_s {
-                log::warn!("activity {} inactive for {}s, destroying", activity_id, dt);
-                enqueue_destroy_evt(
-                    db,
-                    tracker.clone(),
-                    &activity_id,
-                    provider_id,
-                    app_session_id.clone(),
-                )
-                .await;
+            let duration = (Utc::now().timestamp() - usage.timestamp) as f64;
+
+            if duration > unresponsive_limit || duration > inactivity_limit {
+                if state.state.0 != State::Unresponsive {
+                    log::warn!("activity {} unresponsive after {}s", activity_id, duration);
+                    let new_state =
+                        ActivityState::from(StatePair(State::Unresponsive, state.state.1));
+                    prev_state = Some(state);
+                    let _ = tracker
+                        .update_state(activity_id.clone(), State::Unresponsive)
+                        .await;
+                    if let Err(e) = set_persisted_state(&db, &activity_id, new_state).await {
+                        log::error!("cannot update activity {} state: {}", activity_id, e);
+                    }
```

Comment:
It would be good to track it. Note that it is different than counter below for purpose.
And please add counter initialization to 0.
```suggestion
                    }
                    counter!("activity.provider.unresponsive", 1);
```

Reply:
And it would be good to add 
`counter!("activity.provider.responsive-again", 1);`
in the if clause below

Reply:
Counters added.